text/0165-instrumentation-api.md

```diff
@@ -0,0 +1,236 @@
+
+The goals of the instrumentation API are
+
+- Provide a good user experience for authors of library instrumentation that minimizes logic and cognitive
+load.
+
+- Codify quality requirements of good library instrumentation, for example completeness of semantic
+conventions.
+
+- Provide a consistent user experience for users of library instrumentation to ensure similar
+configurability and usability regardless of the library being used.
+
+## Explanation
+
+The instrumentation API is a higher level API built on top of the OpenTelemetry API. The key notion
+is that instrumentation of libraries almost always involves intercepting a library at two points, the
+start and end of an operation, for example an HTTP request or a DB query. This means that the primary
+instrumentation API only needs to provide two methods, `start` and `end` which are called from
+the instrumentation at the beginning and end of the operation. This API is called the `Instrumenter`.
+
+In addition to the operation lifecycle, the other library-specific aspect of instrumentation is
+mapping a library domain object into OpenTelemetry semantic conventions. The instrumentation API
+allows instrumentation to implement `Extractor`s, which map from a request or response object into
+attributes. Extractors enforce a contract, for example by defining an abstract class, which includes
+all the semantic conventions of a given namespace in OpenTelemetry. The library instrumentation just
+implements extractors relevant to the library - for example, an HTTP framework will implement the
+HTTP extractor. This allows all HTTP attributes to be filled without specific knowledge of the
+semantic conventions themselves.
+
+An `InstrumentaterBuilder` stitches together `Instrumenter` and `Extractor`s, as well as consistently
+defining a set of well known configuration knobs for library users. The instrumentation library will
+initialize a builder with defaults registering the implemented `Extractor`s and return it to the
+user of the instrumentation, which can further configure it to their needs.
+
+## Internal details
+
+These details are based on a [prototype](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/9f31a057b6878619157aff93e27fdb15fad89958/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter)
+in Java.
+
+### Instrumenter
+
+The `Instrumenter` is the main entry point for the author of library instrumentation. It has two
+primary methods.
+
+- `Context start(Context parentContext, REQUEST request)`: The method to invoke at the beginning of an
+operation. The parent context, usually the current context, and the library-specific request object
+are passed in. The implementation creates a span, invokes all extractors and request listeners,
+starts the span, and returns a `Context` with the instrumentation state.
+
+- `void end(Context context, REQUEST request, RESPONSE response, Throwable error)`: The method to
```

Comment:
If context could be handled by magic without a javaagent then life would be great :) No ideas on how to improve the general pattern of context management, though at least having 1 method that returns a `Context` and another that is documented to accept it is still a win over OpenTelemetry API I think - I've seen code that creates a context (possibly not parenting correctly) to create a span, then passes in an arbitrary `Context.current()` when ending instead of the instrumentation context.

For dangling spans, I think this is quite language specific. In Java, we can implement a "GC hook" as a span processor to catch spans that get GC'd without `end` being called. Main reason for not having implemented this yet is not knowing what to do with those spans as it probably needs some specification (Zipkin exports the span context (not the span, it's already collected) with with `flush=true` attribute).

Reply:
> If context could be handled by magic without a javaagent then life would be great :)

True. 😄 

> No ideas on how to improve the general pattern of context management, though at least having 1 method that returns a Context and another that is documented to accept it is still a win over OpenTelemetry API I think

I wonder if this is an opportunity to improve the next iteration of the Context / Span APIs themselves?

> For dangling spans, I think this is quite language specific.

This is probably true, and I think we can probably just ignore my comments in this section because of that. Although, maybe mentioning here that languages implementing the Instrumentation API "SHOULD provide an automatic cleanup mechanism for authors", but the value of that may be debatable still.